R/zchunk_LA115.RooftopPV.R

```diff
@@ -30,43 +28,125 @@ module_gcam.usa_LA115.RooftopPV_DISABLED <- function(command, ...) {
+
+    # COMPUTE SUPPLY CURVES
+
+    # Get residential supply curve
+    NREL_Res_PV_supply_curve %>%
+      mutate(p = Relative_Cost * PV_resid_min_LEC_1975,
+             generation = MWh / CONV_BIL_MIL) %>%  # << convert to GWh (like commercial)
+      rename(state_name = State) %>%
+      select(state_name, p, generation) ->
+      L115.pv_sc_res
+
+    # Get commercial supply curve
+    NREL_Com_PV_supply_curve %>%
+      mutate(p = Rel_Cost * PV_comm_min_LEC_1975,
+             generation = GWH) %>%
+      rename(state_name = State) %>%
+      select(state_name, p, generation) ->
+      L115.pv_sc_com
+
+    # Combine supply curves, add state abbrv., and rebase p
+    bind_rows(L115.pv_sc_res, L115.pv_sc_com) %>%
+      left_join_error_no_match(select(states_subregions, state_name, state),
+                               by = "state_name") %>%
+      arrange(state_name) %>%
+      select(-state_name, state, p, generation) %>%
+      mutate(p = p - min(p)) %>%
+      arrange(p) ->
+      # ^^ note: ordering of generation for matching p values will affect supply curve parameters (fitted below).
+      L115.pv_sc
+
+    unique(select(L115.pv_sc, state)) %>% mutate(p = 0, generation = 0) %>%
+      # ^^ adds a 0,0 row to all states for supply curve fitting)
+      bind_rows(L115.pv_sc) %>%
+      arrange(state, p) %>%
+      # ^^ sort for cumulative column calcs to follow
+      group_by(state) %>% mutate(cumul = cumsum(generation),
+                                 percent_cumul = cumul / sum(generation)) ->
+      # ^^ computes cumulative sum (abs. and percent) for each state separately
+      L115.pv_sc_State
+
+    # Get mid-price for each state
+    L115.pv_sc_State %>%
+      filter(percent_cumul > 0.5) %>%
+      # mid-price taken as the first price after the 50th percentile
+      summarise(mid_p = first(p)) -> L115.pv_midPrice
+
+    # Create smoothing functions
+    smooth_res_curve_approx <- function( b, midp, p ) {
+      p_pow_b <- p ^ b
+      f_p <- p_pow_b / ( midp ^ b + p_pow_b )
+    }
+    smooth_res_curve_approx_error <- function( b, midp, supply_curve ) {
+      f_p <- smooth_res_curve_approx( b, midp, supply_curve$p )
+      error <- f_p - supply_curve$percent_cumul
+      crossprod( error, error )
+    }
+    # ^^ computes the smooth renewable resource function at price points...
+    # ... from the actual supply curve, then returns the error between...
+    # ... computed and actual values.
+    get_error_min_b <- function(stateAb) {
+      midPrice <- filter(L115.pv_midPrice, state == stateAb)$mid_p
+      supplyCurve <- filter(ungroup(L115.pv_sc_State), state == stateAb)
+      optimize(f = smooth_res_curve_approx_error, interval = c(1.0, 15.0),
+               midPrice, supplyCurve) %>% unlist
+    }
+    # Get error_min_b for all states using apply (replaces "for loop" structure of legacy code)
+    sapply(unique(L115.pv_sc$state), get_error_min_b) %>% t %>% as.data.frame %>%
```

Comment:
`as_tibble`?

Reply:
the `as.data.frame` is useful here for retaining row names that come out of the sapply function (which later get tranferred into tibble columns).